src/ir_passes/assume.rs

```diff
@@ -0,0 +1,164 @@
+            ast::OrderOp::Gte => ir::CmpOp::gte(lhs, rhs),
+        }))
+    }
+
+    /// Interns a new [ir::Prop] from an [ast::Implication].
+    fn add_implication(
+        &mut self,
+        prop: ast::Implication<ast::Expr>,
+        binding: &Binding<ExprIdx>,
+    ) -> PropIdx {
+        match prop {
+            ast::Implication { guard: None, cons } => {
+                self.add_orderconstraint(cons, binding)
+            }
+            ast::Implication {
+                guard: Some(g),
+                cons,
+            } => self
+                .add_orderconstraint(g, binding)
+                .implies(self.add_orderconstraint(cons, binding), self),
+        }
+    }
+}
+
+impl Assume {
+    /// Checks a proposition for whether it matches the form `#l = f(#r)` for some custom function `f`. Additionally recurses on `&` chains.
+    /// Generates the assumptions associated with each [ast::UnFn] and returns a list of [ir::Prop]s for each.
+    /// TODO: Implement assumption generation for functions taking more than one argument.
+    fn prop(p: ir::PropIdx, comp: &mut ir::Component) -> Vec<PropIdx> {
+        let p = comp.get(p);
+        match p {
+            ir::Prop::Cmp(ir::CmpOp {
+                op: ir::Cmp::Eq,
+                lhs,
+                rhs,
+            }) => {
+                // Matches over the cases `op(args) = rhs` and `lhs = op(args)` to
+                // define the `op`, `left`, and `right` for the equivalent equation `left = op(right)`
+                if let Some((op, left, right)) = match (
+                    comp.get(*lhs),
+                    comp.get(*rhs),
+                ) {
+                    (ir::Expr::Fn { op, args }, _) => {
+                        assert!(
+                            args.len() == 1,
+                            "Currently Unimplemented: {} requires {} arguments, automatic assumptions only implemented for single argument functions.",
+                            op, args.len()
+                        );
+                        Some((op, *rhs, args[0]))
+                    }
+                    (_, ir::Expr::Fn { op, args }) => {
+                        assert!(
+                            args.len() == 1,
+                            "Currently Unimplemented: {} requires {} arguments, automatic assumptions only implemented for single argument functions.",
+                            op, args.len()
+                        );
+                        Some((op, *lhs, args[0]))
+                    }
+                    _ => None,
+                } {
+                    log::debug!("Generating default assumptions for {p}");
+                    // Creates the binding for the left and right [ExprIdx]s
+                    let binding = Binding::new(vec![
+                        (FnAssume::left(), left),
+                        (FnAssume::right(), right),
+                    ]);
+                    FnAssume::from(op.clone())
```

Comment:
I see, the `ast` dependence comes from here because `FnAssume` works on `ast` objects. It would be annoying but I think it would be worth duplicating the code since we're going to delete the passes from the non-IR pass once the transition is complete.